eth/executionclient/multi_client.go

```diff
@@ -0,0 +1,476 @@
+	if err != nil {
+		return nil, err
+	}
+
+	return res.([]ethtypes.Log), nil
+}
+
+func (ec *MultiClient) Filterer() (*contract.ContractFilterer, error) {
+	return contract.NewContractFilterer(ec.contractAddress, ec)
+}
+
+func (ec *MultiClient) ChainID(_ context.Context) (*big.Int, error) {
+	return ec.chainID, nil
+}
+
+func (ec *MultiClient) Close() error {
+	close(ec.closed)
+
+	var multiErr error
+	for i, client := range ec.clients {
+		if err := client.Close(); err != nil {
+			ec.logger.Debug("Failed to close client", zap.String("address", ec.nodeAddrs[i]), zap.Error(err))
+			multiErr = errors.Join(multiErr, err)
+		}
+	}
+
+	return multiErr
+}
+
+func (ec *MultiClient) call(ctx context.Context, f func(client SingleClientProvider) (any, error)) (any, error) {
+	if len(ec.clients) == 1 {
+		return f(ec.clients[0])
+	}
+
+	for i := 0; i < len(ec.clients); i++ {
+		ec.currClientMu.Lock()
+		currentIdx := ec.currClientIdx
+		ec.currClientMu.Unlock()
+
+		client := ec.clients[currentIdx]
+
+		ec.logger.Debug("checking client healthiness",
+			zap.Int("index", currentIdx),
+			zap.String("method", methodFromContext(ctx)),
+			zap.String("addr", ec.nodeAddrs[currentIdx]))
+
+		// Make sure this client is healthy, this shouldn't cause too many requests because the result is cached.
+		// TODO: Make sure the allowed tolerance doesn't cause issues in log streaming.
+		if err := client.Healthy(ctx); err != nil {
+			ec.logger.Warn("client is not healthy, using another one",
+				zap.Int("index", currentIdx),
+				zap.String("method", methodFromContext(ctx)),
+				zap.String("addr", ec.nodeAddrs[currentIdx]))
+
+			ec.useNextClient(ctx, currentIdx, client)
+
+			continue
+		}
+
+		ec.logger.Debug("calling client",
+			zap.Int("index", currentIdx),
+			zap.String("method", methodFromContext(ctx)),
+			zap.String("addr", ec.nodeAddrs[currentIdx]))
+
+		v, err := f(client)
+		if errors.Is(err, ErrClosed) || errors.Is(err, context.Canceled) {
+			ec.logger.Debug("received graceful closure from client", zap.Error(err))
+			return v, err
+		}
+
+		if err != nil {
+			ec.logger.Error("call failed, trying another client",
+				zap.String("method", methodFromContext(ctx)),
+				zap.String("addr", ec.nodeAddrs[currentIdx]),
+				zap.Error(err))
+
+			ec.useNextClient(ctx, currentIdx, client)
+
+			continue
+		}
+
+		ec.logger.Debug("call succeeded, returning value",
+			zap.String("method", methodFromContext(ctx)),
+			zap.String("addr", ec.nodeAddrs[currentIdx]))
+		return v, nil
+	}
+
+	return nil, fmt.Errorf("all clients returned an error")
+}
+
+func (ec *MultiClient) useNextClient(ctx context.Context, currentIdx int, client SingleClientProvider) {
+	ec.currClientMu.Lock()
+	idx := ec.currClientIdx
+	// The index might have already changed if a parallel request failed.
+	if idx == currentIdx {
+		ec.currClientIdx = (ec.currClientIdx + 1) % len(ec.clients)
+	}
+	ec.currClientMu.Unlock()
+
+	go client.reconnect(ctx)
```

Comment:
this used to happen only in streaming, so we should either do the same, or rely on ethclient's built-in reconnection (requires properly testing it)